### PR TITLE
Working Global Projections with basic concepts

### PR DESCRIPTION
**Pull request by Sandreu**
#15 
As I said here is the basic implementation... Just to have something to talk about...
Main todos are:
- Reorder events from all context before playing.
- Wait a context creation if it doesn't exist yet ?

Regards

### CODE REVIEW

**Comment by johannesbecker** [comment]
I don't get yet why the creation of the "GlobalProjectionsContext" is necessary. Could you explain why?

**Comment by Sandreu** [comment]
I created a globalProjectionContext to be able to listen change events from another service, and also to avoid stores creation code duplications...

The projection service is very dependent of context... I just wanted to make it work, then talk about necessary refactoring. 

I think that perhaps eventric should be a pubsub and publish events as context:ContextName:initialized or global projection changes...

**Comment by johannesbecker** [comment]
Yeah, I think it's a good basis to iterate on. Is there still anything to clean up before we merge?

File: src/global_projection/global_projection_service.coffee [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        contextList = @_getContextsFromProjection projection
+        for contextName in contextList
+          subContext = @_eventric.getContext contextName
+          if not subContext then throw new Error("Context #{contextName} doesn't exists")
+          
+          contextEvents = @_getEventsForContext projection, contextName
+          subProjection = {}
+          
+          for eventName in contextEvents
+            subProjection['handle'+eventName] = this._eventHandler projection, contextName, eventName
```
Comment (johannesbecker) [inline src/global_projection/global_projection_service.coffee]: this. should actually be @ here

**Comment by Sandreu** [comment]
Yes I saw 2 console.log in specs :)

**Comment by Sandreu** [comment]
I don't understand your diff comment... :s

**Comment by johannesbecker** [comment]
`this._eventHandler` -> `@_eventHandler`

**Comment by Sandreu** [comment]
Oh yeah... :)

**Comment by alexlawrence** [comment]
If `subContext = @_eventric.getContext contextName` would be `subContext = @_eventric.remote contextName` would this work for remotes also?

**Comment by Sandreu** [comment]
Good question. I didn't look how remotes works, but I don't know if it's possible to create a projection to a remote context... I'm currently working on my SharedProjections, when it works, I come back to global projection to order events playback and to take a look on remotes.

**Comment by Sandreu** [comment]
Hello,

I just found while playing with GlobalProjections that domainEvents can be lost if emitted while projection's `_applyDomainEventsFromStoreToProjection`.

That's not the case if all projections are added before context's initialization, but if I initialize a context then add a projection and manually initialize it, events emitted while `projection->_applyDomainEventsFromStoreToProjection` are lost.

I assume that it's not an issue and projections have to be added before context initialization (perhaps this could be a feature ?). Then I assume with global projections that no events are emitted in contexts while Global projection is initializing.

> If subContext = @_eventric.getContext contextName would be subContext = @_eventric.remote contextName would this work for remotes also?

Not yet (because of `context._initialized`) but it will :)

**Comment by johannesbecker** [comment]
A thing that crossed my mind is the API "addGlobalProjection" - does it have to be a seperate thing? Couldn't existing Projections just be able to handle "fromContext_" syntax out of the box? Or do I miss something important?

Besides that @alexlawrence is right, they should definitely work on Remotes too.

**Comment by Sandreu** [comment]
lol, I was asking myself the same question :)... as a global projection is attached to a globalContext, why context's projections couldn't have their own `fromContext_handlers` I go for it..

**Comment by johannesbecker** [comment]
> I assume that it's not an issue and projections have to be added before context initialization (perhaps this could be a feature ?)

It's already a feature on the Remotes (see https://github.com/efacilitation/eventric/blob/master/src/remote/remote.coffee#L278-L287). Actually porting this to the Context should be just copy&paste the initializeProjection function.

**Comment by Sandreu** [comment]
[https://github.com/efacilitation/eventric/blob/master/src/context/context.coffee#L527-L534](https://github.com/efacilitation/eventric/blob/master/src/context/context.coffee#L527-L534)

This is too on contexts... But (even on remotes) if you generate an event while projection is initializing, your projection won't listen on this event... as events is an array [here](https://github.com/efacilitation/eventric/blob/master/src/projection/projection.coffee#L149-L161) and [subscription is after playback](https://github.com/efacilitation/eventric/blob/master/src/projection/projection.coffee#L57-L61)

**Comment by johannesbecker** [comment]
It's not the same thing.

**Comment by Sandreu** [comment]
Well I don't get the difference...

 and why

> If you generate an event while projection is initializing, your projection won't listen on this event... as events is an array [here](https://github.com/efacilitation/eventric/blob/master/src/projection/projection.coffee#L149-L161) and [subscription is after playback](https://github.com/efacilitation/eventric/blob/master/src/projection/projection.coffee#L57-L61)

is not true with remotes...?

And initialize a projection without name isn't problematic with stores ?

**Comment by johannesbecker** [comment]
> is not true with remotes...?

The issue #19 exists with remotes too.


> And initialize a projection without name isn't problematic with stores ?

We're currently using the directly initialized projection only on Remotes without stores (as angular directives). Could be that they currently not seamlessly work with projection stores.

**Comment by Sandreu** [comment]
See #20

**Pull request closed without merging**